cmd-ref: plots: flexible plots docs

pared opened the pull request:
Related to: iterative/dvc#7477 (iterative/dvc#7086)

Docs for flexible plots.
Added `templates` command reference.
Partially solves #2956. - It seems to me that there is no point in differentiating between metrics/plots notation. Plots are no longer special types of output. Main point of the `plots` is to provide visualization. Visualizations can be created from any supported type of data, hence we only need to mention that there exists special type of `output` (`plots`) That can help with visualizations.

Do not merge before: iterative/dvc#7477 is merged.

In review app: https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/command-reference/plots

UPDATE: Jump down to https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1045840235.

shcheklein commented on content/docs/command-reference/plots/templates.md:
```diff
@@ -0,0 +1,95 @@
+## plot templates
+
+Dump plot templates. Dumps the templates to `.dvc/plots` by default. Note that
```
If we just want to say that it dumps plot templates - the first sentence is enough (no need to include defaults, etc). It's better to come up with a sentence that explains the motivation / use case for this command instead.

pared replied:
Modified.

shcheklein commented on content/docs/command-reference/plots/templates.md:
````diff
@@ -0,0 +1,95 @@
+## Synopsis
+
+```usage
+usage: dvc plots templates [-h] [-q | -v] [-o <path>] [{simple,linear,confusion,confusion_normalized,scatter,smooth}]
````
looks like it's not scalable, making a list of all available names. Usually you would have a command that can list them. Or name it a `<template>` and in the details list the available plots.

pared replied:
We have created this command intentionally this way to not avoid unnecessary subcommands.
https://github.com/iterative/dvc/pull/7108
If we want to modify it we should probably open a new issue in dvc and get back to it.

shcheklein replied:
yes, it's more about this command's interface in this case. Feels like this list `{simple,linear,confusion,confusion_normalized,scatter,smooth}` will be growing / changing. Why do we have to specify it on the parser level, it's not very common to limit that large number of targets there.

shcheklein commented on content/docs/command-reference/plots/templates.md:
```diff
@@ -0,0 +1,95 @@
+- `confusion_normalized` - confusion matrix with values normalized to <0, 1>
+  range.
+
+### Custom templates
```
It's not exactly clear why we have it here tbh. It's related to this command except its name/

pared replied:
It seemed to me that explaining how templates work makes most sense in a command description directly related to it. I am not so sure now. I am also not sure it should be in index. Maybe we need a separate `user-guide` entry for `customizning plots`?

shcheklein replied:
Okay,I see that the existing description got modified (we are explaining why do we need this command - not a regular thing to do, which is a bit of a bad sign). I think it's fined now.

shcheklein commented on content/docs/command-reference/plots/index.md:
```diff
@@ -1,57 +1,56 @@
+    show                Generate plots from target files or plots definitions from `dvc.yaml` file.
+    diff                Show multiple versions of plot data by plotting them in a single image.
+    modify              Modify display properties of data-series plot outputs (has no effect on image-type plots).
+    templates           Write built-in plots templates to a directory (.dvc/plots by default).
```
would be better to have `dvc plots templates` (show available templates, including custom?), `dvc plots template --dump <template-name>` dumps it. 

(feels like dumping a template is an aux command compared to show / diff  and even modify)

pared replied:
We can discuss and modify the usage, but I would prefer to avoid it in this PR.

shcheklein commented on content/docs/command-reference/plots/index.md:
````diff
@@ -1,57 +1,56 @@
+  -v, --verbose         Be verbose.
 ```
 
-## Types of metrics
````
do we want to move it somewhere? (not sure if we have a good place)

jorgeorpinel replied:
This is in https://dvc.org/doc/command-reference/metrics#types-of-metrics now. At least link from `plots` ref(s), yeah.

pared replied:
As I mentioned in the [issue description](https://github.com/iterative/dvc.org/pull/3691#issue-1283821901) I don't think we should be trying to explain plots in terms of metrics types. `dvc plots` is a visualization that can be created from any type of viable data. Yes, we have a convenient output type that lets DVC pick up plots without defining them but I believe saying that its special type of metrics makes users think about it in a particular way that is unnecessary.

shcheklein replied:
Agreed, and as I mentioned I'm not sure we have a good place for this.

shcheklein replied:
@jorgeorpinel let's drop it in the second place as well?

shcheklein commented on content/docs/command-reference/plots/index.md:
```diff
@@ -280,7 +250,157 @@ file:///Users/usr/src/dvc_plots/index.html
+      train_classes.csv: predicted_class
+      test_classes.csv: predicted_class
+    title: Compare test vs train confusion matrix
+    template: confusion
```
how does diff look like in this case (unmergable plots)? can you show an example please? just trying to understand the consequences for VS Code / Studio

shcheklein replied:
Got it. I would say it's confusing to see `rev` as a title a bit. Not related to this PR obviously.

I think it's fine to have a simple case for these plots. I don't it's required or something.

pared replied:
Yep, I will have to fix that. Regretfully, because in the beginning, I did not convert `rev` into an anchor, this field will have to stay with us for some time. I should probably be able to remove it from the subtitle though.

dberenbaum replied:
Do we think it's useful to include a diff example here (doesn't need to block the PR)?

shcheklein commented on content/docs/command-reference/plots/index.md:
```diff
@@ -280,7 +250,157 @@ file:///Users/usr/src/dvc_plots/index.html
+# dvc.yaml
+stages:
+  train:
+    cmd: echo "Hello world"
```
make it more realistic? or remove completely (we don't anything like this in all the other cases)

shcheklein replied:
it makes reproducible though ...

make it "Training a model ... " at least?

jorgeorpinel replied:
I think in ref examples realistic > reproducible. We don't expect readers to run the examples as in a tutorial right? They're mostly to see them, maybe copy the code but change it to fit their project.

pared replied:
Modified.

shcheklein reviewed (changes requested):
How does it affect the `plots modify`? can I pass to it some extra arguments?

shcheklein commented on content/docs/command-reference/plots/templates.md:
```diff
@@ -0,0 +1,95 @@
+
+- `<DVC_METRIC_X_LABEL>` (optional) - field name to display as the X axis label
+
+- `<DVC_METRIC_Y_LABEL>` (optional) - field name to display as the X axis label
```
X -> Y

pared replied:
Modified.

shcheklein commented on content/docs/sidebar.json:
```diff
@@ -388,7 +388,8 @@
             "slug": "modify"
-          }
+          },
+          { "label": "plots templates", "slug": "templates" }
```
minor: style looks different?

jorgeorpinel replied:
Yeah I'm surprised the linters didn't complain. I guess we could use compressed lines like this but... for now being consistent (committing):
```suggestion
          {
            "label": "plots templates",
            "slug": "templates"
          }
```

shcheklein reviewed (changes requested):
It feels we need somewhere a better description of the possible structure of the `plots` yaml section (table with examples)?

do we need to update the schema file for the dvc.yaml?

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -1,57 +1,56 @@
-    show         Generate plot from a metrics file.
-    diff         Plot differences in metrics between commits.
-    modify       Modify display properties of data-series plots (has no effect on image-type plots).
+  {show,diff,modify,templates}
```
Should be this here I think?
```suggestion
  COMMAND
```

jorgeorpinel replied:
Should be this here I think?
```suggestion
  COMMAND
```

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -1,57 +1,56 @@
-    diff         Plot differences in metrics between commits.
-    modify       Modify display properties of data-series plots (has no effect on image-type plots).
+  {show,diff,modify,templates}
+                        Use `dvc plots CMD --help` to display command-specific help.
```
No need for this line in the refs.
```suggestion
```

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -1,57 +1,56 @@
+    diff                Show multiple versions of plot data by plotting them in a single image.
+    modify              Modify display properties of data-series plot outputs (has no effect on image-type plots).
+    templates           Write built-in plots templates to a directory (.dvc/plots by default).
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -q, --quiet           Be quiet.
+  -v, --verbose         Be verbose.
```
Or any of these...
```suggestion
```

jorgeorpinel reviewed (commented):
Stylistic changes to Synopsis... please apply to content/docs/command-reference/plots/templates.md as well 🙂

pared commented:
> How does it affect the plots modify? can I pass to it some extra arguments?

For now `plots modify` work as they used to - for `plot` types of outputs. I didn't change the behavior not to bloat the existing change. There is also a question of whether `modify` should allow creating a new plot in `dvc.yaml`

pared commented:
> do we need to update the schema file for the dvc.yaml?

@shcheklein 
You mean to include `plots` in `doc/user-guide/project-structure/pipelines-files`?

pared commented:
> Stylistic changes to Synopsis... please apply to content/docs/command-reference/plots/templates.md as well

Modified the `templates.md` synopsis

shcheklein commented on content/docs/command-reference/plots/templates.md:
```diff
@@ -0,0 +1,97 @@
+usage: dvc plots templates [-h] [-q | -v] [-o <path>] [{simple,linear,confusion,confusion_normalized,scatter,smooth}]
+
+positional arguments:
+  TEMPLATE
```
I think we are using lower case names?

jorgeorpinel replied:
Moved to https://github.com/iterative/dvc.org/pull/3697#pullrequestreview-1024032767

pared commented:
@dberenbaum I think that we can split templates docs, thought I don't see why should we split #3295 and #2956 - the change in way of thinking about plots is part of #3295 so I think those two should be part of a single change.

pared commented:
@dberenbaum @jorgeorpinel @shcheklein 
Created #3697 for `templates`.
Not editing the current state of templates as it is addressed there.

pared commented:
Added some info to `pipelines-files.md`

shcheklein commented on content/docs/command-reference/plots/index.md:
````diff
@@ -96,7 +87,44 @@ names in the `train` array below:
+plots: ...
+```
+
+## Plot outputs
````
By reading this I don't understand how plot outputs are not related to the global "plots" section. Could we clarify this?

dberenbaum replied:
See also https://github.com/iterative/example-repos-dev/pull/117#issuecomment-1175197428

shcheklein commented on content/docs/user-guide/project-structure/pipelines-files.md:
```diff
@@ -114,6 +114,35 @@ stages:
+    metrics:
+      - accuracy.txt:
+          cache: false
+plots:
```
are those still supported per stage?

pared replied:
You mean `plots outputs`? Yes, they are supported

jorgeorpinel replied:
OK but maybe then the `outs` row (**Stage entries** table) should mention these can become viable for visualization via [top-level plot definitions](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/user-guide/project-structure/pipelines-files#top-level-plot-definitions).

shcheklein commented on content/docs/command-reference/plots/index.md:
```diff
@@ -280,7 +304,157 @@ file:///Users/usr/src/dvc_plots/index.html
+
+![](/img/plots_show_confusion_normalized.svg)
+
+## Example: simple plot definition
```
My feeling that we should create a large table with three-four columns - data, template, spec (dvc.yaml), image - and put a lot of examples there - from some basic regular things, to more advanced. So that people can quickly start, search for it, copy paste, etc

dberenbaum replied:
Some more organization would definitely be useful, and more examples would be great.

For now, could we at least try to break it up so it's not one huge list of examples? Maybe have different subsections. It seems like there's a bunch of path-based examples, followed by dvc.yaml examples, and then finally a custom html template example.

dberenbaum commented on content/docs/user-guide/project-structure/pipelines-files.md:
```diff
@@ -114,6 +114,35 @@ stages:
 The commands in `dvc metrics` and `dvc plots` help you display and compare
 metrics and plots.
 
+## Plots definitions
```
Thanks for adding this section @pared. "Pipelines files" is no longer a good descriptor for this page if we include it, but I'm not sure there is a better place for it now, and I would say let's not let it block this PR for now.

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +87,44 @@ names in the `train` array below:
+
+## Plots definitions
+
+Plots defined in `dvc.yaml` are especially useful when users want to compare
```
```suggestion
Plots defined in `dvc.yaml` under the `plots` key are especially useful when users want to compare
```

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +87,44 @@ names in the `train` array below:
+## Plots definitions
+
+Plots defined in `dvc.yaml` are especially useful when users want to compare
+data from differend data sources residing on the same version of the project.
```
```suggestion
data from different data sources residing on the same version of the project.
```

dberenbaum commented on content/docs/user-guide/project-structure/pipelines-files.md:
````diff
@@ -114,6 +114,35 @@ stages:
+    y: tpr
+```
+
+Note that in this example we don't have to specify `auc.json` as a plot output.
````
It might help to explain why this is useful and why there are two different ways to define plots. Maybe provide an example that uses multiple files or otherwise illustrates its use?

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +87,44 @@ names in the `train` array below:
+<abbr>outputs</abbr> as plot or define plot configuration inside `dvc.yaml`
+under `plots` key.
+
+## Plots definitions
```
Suggestion: Can the headers and descriptions better explain the two different ways to define plots? Maybe introduce some terminology like "Standalone Plots" and "Stage Plots"?

dberenbaum replied:
Also, should both this and the next header be subheadings of "Defining a plot"?

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +87,44 @@ names in the `train` array below:
+When using `dvc run` or `dvc stage add`, instead of using
+`--outs/--outs-no-cache` particular outputs can be marked with
+`--plots/--plots-no-cache`. This will tell DVC that they are intended for
+visualizations. This special type of outputs might come in hand if users want to
+visually compare experiments results with other experiments versions. For
+example, comparing new experiment with the baseline version of the project.
```
Why is this info under the section specific to plots outputs?

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -280,7 +304,157 @@ file:///Users/usr/src/dvc_plots/index.html
+
+![](/img/plots_show_spec_simple_custom.svg)
+
+## Example: multiple data-series plot definition:
```
```suggestion
## Example: multiple data-series plot definition
```

dberenbaum commented on content/docs/command-reference/plots/index.md:
````diff
@@ -280,7 +304,157 @@ file:///Users/usr/src/dvc_plots/index.html
+      training_data.csv: [test_loss, train_loss]
+    title: Compare loss training versus test
+```
+
````
`dvc plots show`?

dberenbaum commented on content/docs/command-reference/plots/index.md:
````diff
@@ -280,7 +304,157 @@ file:///Users/usr/src/dvc_plots/index.html
+    x_label: Actual class
+    y_label: Predicted class
+```
+
````
`dvc plots show`?

dberenbaum commented on content/docs/user-guide/project-structure/pipelines-files.md:
```diff
@@ -114,6 +114,35 @@ stages:
+Note that in this example we don't have to specify `auc.json` as a plot output.
+Plots definitions accept any file from repository. For more examples how to
+create plots definitions refer to the
+[examples](/doc/command-reference/plots#example-simple-plot-definition).
```
We might need more explanation of the syntax either here or in the cmd ref. For example (I think these are correct?):

* The top-level key can be either the path to a file to be used as the data source, or an arbitrary unique name. It will be used as the plot title unless another title is set.
* `y` can have multiple entries to plot multiple data series. Each `y` entry can also be a key-value pair of the file path and field name.

Alternatively, this might fit as comments within the dvc.yaml examples, but it might help to explain a bit even if we want to focus on examples.

pared commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +87,44 @@ names in the `train` array below:
+
+### Syntax
+
+In order to define the plot users need to provide data and configuration for the
```
No

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
````diff
@@ -12,46 +12,37 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify} ...
+    modify       Modify display properties of data-series plot outputs (has no effect on image-type plots).
 ```
 
-## Types of metrics
-
-DVC has two concepts for metrics, that represent different results of machine
-learning training or data processing:
````
Should we also kill or rephrase this section in the `metrics` ref. since we're now explaining `plots` as a data-related command? As well as other `plots` links from that ref.

See https://dvc.org/doc/command-reference/metrics#types-of-metrics and 3rd paragraph in the **Description** below that.

pared replied:
I agree. Should we just remove it? It seems to me that it doesn't make sense to keep this paragraph.

jorgeorpinel replied:
> Should we just remove it?

Yes please @pared

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -12,46 +12,37 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify} ...
-DVC provides a set of commands to visualize certain metrics of machine learning
-experiments as plots. Usual plot examples are AUC curves, loss functions,
-confusion matrices, among others.
+DVC provides a set of commands to visualize data produced by machine learning
+experiments. Usual plot examples are AUC curves, loss functions, confusion
+matrices, among others.
```
I agree to present `plots` mainly as a data-centric command instead of another type of "metric" (which was always a little confusing). But one use case is still to use `plots` to visualize "2D" pipeline metrics right? So I think we can elaborate a little in this description:

```suggestion
DVC provides a set of commands to visualize data produced by machine learning
projects. Usual plots include AUC curves, loss functions, or confusion
matrices, for example.

Plots are a great alternative to `dvc metrics` when working with multi-dimensional
performance data. They also help you present and compare [experiments] effectively.
```

☝🏼 `[experiments]` link missing

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -12,46 +12,37 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify} ...
-This type of metrics files are created by users, or generated by user data
-processing code, and can be defined in `dvc.yaml` (`plots` field) for tracking
-(optional).
+This type of data is created by users, or generated by user data processing
+code.
 
```
No longer needed (redundant) I think.
```suggestion
```

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -12,46 +12,37 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify} ...
-DVC can work with two types of plots files:
+## Types of plots
+
+`dvc plots` is able to visualize two types of data:
```
I'd roll this change back, or at least avoid adding more structure. The **Description** should not need more H2/3s. If it does it's a sign that we need a new User Guide on top of this change...

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -12,46 +12,37 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify} ...
 
-DVC generates plots as static HTML webpages that can be open with a web browser.
-They can also be saved as SVG or PNG image files from the browser.
+DVC generates visualizations as static HTML webpages that can be open with a web
+browser. They can also be saved as SVG or PNG image files from the browser.
```
Rolling back, it was clearer before 🙂 
```suggestion
DVC generates plots as static HTML webpages that can be open with a web browser.
They can also be saved as SVG or PNG image files from the browser.
```

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -12,46 +12,37 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify} ...
+rendering (declarative JSON grammar for defining graphics). Images are rendered
+using `<img>` tags directly.
 
 ## Supported file formats
```
Hmmm... I see we already have all [this structure](https://dvc.org/doc/command-reference/plots#supported-file-formats) in. I'll make an issue about flattening this/ extracting to some guide ⏳  For now most of these should probably be H3s (contained under the **Description**)?

jorgeorpinel replied:
> I'll make an issue about flattening this/ extracting to some guide 

A closely-related epic already exists so I just updated that one: https://github.com/iterative/dvc.org/issues/2572

pared replied:
> Hmmm... I see we already have all [this structure](https://dvc.org/doc/command-reference/plots#supported-file-formats) in. I'll make an issue about flattening this/ extracting to some guide ⏳ For now most of these should probably be H3s (contained under the Description)?

You mean `Definig a plot too`?

jorgeorpinel reviewed (commented):
Review on general changes to `plots` parent ref Description.

> I see the reviews for this and #3697 are getting a bit heavy. Lmk if you need more hands-on help.

dberenbaum commented on content/docs/command-reference/plots/index.md:
````diff
@@ -96,7 +84,143 @@ names in the `train` array below:
 ```
 
-## Plot templates (data series only)
+## Definining a plot
````
"Configuring a plot"? WDYT?

It might help here to explain that you can specify a data file and configure the plot in the plots commands. Adding a plot to dvc.yaml saves you from having to repeat it and allows you to consistently view a specified set of plots. dvc.yaml can also persist your plots configuration and enables you to combine data from multiple files into one plot.

@jorgeorpinel Any thoughts/ideas on wording?

pared replied:
> enables you to combine data from multiple files into one plot.

This is pretty much in the next paragraph... should I add it here too?

dberenbaum replied:
Sorry @pared, my focus on this comment was on the benefits of putting plots configuration into `dvc.yaml`, regardless of whether they are standalone or stage plots.

On second thought, I think we should include `dvc.yaml` in the section title ("Configuring in `dvc.yaml`"?). Although plots may be configured from the CLI, neither the data sources nor the configuration get saved, and only a subset of plots configuration is actually supported in the current CLI (for example, plots with multiple data sources can't be configured in the CLl).

WDYT?

jorgeorpinel replied:
> "Configuring a plot"? WDYT?

Sorry for the delay @dberenbaum (I didn't want to add yet another reviewer... but here we are).

Actually I think "defining" is better. But I'm about to make a suggestion on what to do with all these very long blocks of new content (may need another PR split TBH).

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +84,143 @@ names in the `train` array below:
+
+### Standalone plots
+
+Plots defined in `dvc.yaml` under the `plots` key are especially useful when
+users want to compare data from different data sources residing on the same
+version of the project. For example, comparing training versus test results on
```
```suggestion
Plots can be defined in `dvc.yaml` under the `plots` key. Unlike
[stage plots](#stage-plots), these are especially useful when
users want to compare data from different data sources residing on the same
version of the project. For example, comparing training versus test results on
```

It might be better to split this into two sentences. Explain what standalone plots are, then explain their unique benefit in the next sentence (and reference stage plots since it's not otherwise clear what you are comparing them to).

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +84,143 @@ names in the `train` array below:
+`test_logs.csv` against the `epoch`. Note that both files have to have `epoch`
+field.
+
+#### Avialable configuration fields
```
```suggestion
#### Available configuration fields
```

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +84,143 @@ names in the `train` array below:
+  be the path to the data source. String, or list elements will be the names of
+  data columns or fields withing the source file. If this field is a dictionary,
+  it is assumed that its keys are paths to data sources. The values have to be
+  either strings or lists, and are treted as column(s)/field(s) within
```
```suggestion
  either strings or lists, and are treated as column(s)/field(s) within
```

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +84,143 @@ names in the `train` array below:
+`--plots/--plots-no-cache`. This will tell DVC that they are intended for
+visualizations.
+
+Upon running `dvc plots show/diff` DVC will collect plot outputs alongside the
+definitions and display them conforming to their configuration. Note, that if
+there are plot outputs in the project and they are also used in some plots
+definitions, DVC will create separate rendering for the plot outputs and all
+definitions using them.
```
Let's consistently use one set of terms for the two types of plots in dvc.yaml. The headings use "standalone plots" and "stage plots," but most of the text uses "plot outputs" and "definitions." Let's pick one and apply it everywhere.

pared replied:
Refactored.
Follow-up changes for help:
https://github.com/iterative/dvc/pull/7972

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +84,143 @@ names in the `train` array below:
+[examples](/doc/command-reference/plots#example-simple-plot-definition) for more
+syntax insight.
+
+### Stage plots
```
WDYT about moving this above Syntax and then mentioning in Syntax when some options are specific to standalone plots definitions?

pared replied:
Syntax is still part of plot configuration, so I would go with h3, to make it under `Configuring a plot`

shcheklein commented on content/docs/user-guide/project-structure/pipelines-files.md:
````diff
@@ -408,6 +408,35 @@ validation and auto-completion.
+    y: tpr
+```
+
+Note that in this example we don't have to specify `auc.json` as a plot output.
````
Q - can it be DVC cached?

pared replied:
Yes

shcheklein commented on content/docs/user-guide/project-structure/pipelines-files.md:
```diff
@@ -408,6 +408,35 @@ validation and auto-completion.
 
+## Plots definitions
+
+The list of plots contains one or more user-defined
```
@dberenbaum @pared @daavoo what is the usual case that is expected for this global "plots" declaration? Is it primarily about bootstrapping the repo initially that doesn't have any stages?) should we clarify this? make example simpler below?

from the example it's not clear what it the benefit of having it outside the pipeline ..

pared replied:
I would say it's explained in `plots` docs. I don't think it should be a part of `dvc.yaml` docs - it only informs that such thing as `plots` exists.

dberenbaum replied:
This is what I meant in https://github.com/iterative/dvc.org/pull/3691#discussion_r908927506. I agree with @shcheklein that we should provide some brief explanation for why there are `plots` both here and under `stages` (or the examples should make this clear) since this doc is the holistic explanation of the `dvc.yaml` structure.

> Is it primarily about bootstrapping the repo initially that doesn't have any stages?

This is actually not possible yet ☹️ and is a follow-up action. The `dvc.yaml` must already have `stages` for now.

The benefit of the global `plots` is that if your plot sources multiple outputs, there's no way to configure it in your stage outputs.

daavoo commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +84,145 @@ names in the `train` array below:
+
+### Stage plots
+
+When using `dvc run` or `dvc stage add`, instead of using
```
```suggestion
When using `dvc stage add`, instead of using
```

daavoo commented on content/docs/command-reference/plots/index.md:
```diff
@@ -96,7 +84,145 @@ names in the `train` array below:
+visualizations.
+
+Upon running `dvc plots show/diff` DVC will collect stage plots alongside the
+standalone plots and display them conforming to their configuration. Note, that
+if there are stage plots in the project and they are also used in some plots
+standalone plots, DVC will create separate rendering for the stage plots and all
+standalone plots using them.
```
I find this phrase a little confusing. Do we want to explain something like that the actual data files can be reused in the standalone plots and the stage plots?

jorgeorpinel replied:
Yeah the terminology is confusing. Issues I see:

**Configuring a plot**

Why is is separate from `plots modify` and **Plot templates**? Those are 2 actual ways to configure plots (already slightly confusing that there's 2 plot configs, let's avoid a 3rd). I think we're talking about "defining plots" here...

**Stand-alone plots**

They're linked to a stage via an `output` name though, right? So why stand-alone? I think this is about reusable, pipeline-level definitions.

**Stage plots**

Sounds like we're adding another dimension to the concept of "plot types" (dimension 1 being data-series vs. images) which may be confusing. And again, even pipeline-level plot definitions are linked to stages via an `output` name, so the term isn't accurate.

> Will try to post suggestions once I get a better understanding of the big picture...

jorgeorpinel replied:
OK here's what I'm thinking: This ref is now way too long. And even when clearly [we need to make a proper guide](https://github.com/iterative/dvc.org/issues/2572), a lot of the detail here can be compacated and reorganized. Specific plan of action I recommend:

**Configuring a plot** / **Stand-alone plots**/ **Stage plots**

Make this a paragraph or 2 in the existing **Description** of this page (ideally no H3 sections, or make a single **Defining plots** section if absolutely necessary). It probably replaces **Supported file formats** since that info is included in the plot definition and syntax blocks.

Terminology-wise, I'd go with descriptive phrasing like "plots defined in stages", "found in the `plots` field of `dvc.yaml`", etc. instead of coming up with special terms. From the user-perspective all plots are the same once rendered.

**Syntax (standalone plots only)**

Move all of this into [**Standalone plots configurations**](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/user-guide/project-structure/pipelines-files#standalone-plots-configurations) (in the dvc.yaml schema guide) and link from this page as needed.

> I'd rename that section to **Reusable plot definitions** (will leave a specific suggestion ⏳)

Try to summarize as much as possible. Use tables if needed. I can jump in to help with that once the content is in the right place. Thanks

daavoo reviewed (approved):
Minor comments regarding structure but I guess its not for this P.R. but for a follow-up in order to extract most of this info into some `user-guide`

jorgeorpinel commented on content/docs/command-reference/plots/modify.md:
```diff
@@ -1,10 +1,11 @@
 > ⚠️ Note that this command can modify only data-series plots. It has no effect
-> on image-type plots. See
-> [Types of metrics](/doc/command-reference/plots#types-of-metrics).
+> on image-type plots and standalone plots. See
+> [Types of plots](/doc/command-reference/plots#description).
```
Users can click previous links to find out about general info on plots, no need to update this link since that section is gone (we can remove it). Committing
```suggestion
> on image-type plots and standalone plots.
```

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -12,30 +12,18 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify} ...
-    show         Generate plot from a metrics file.
-    diff         Plot differences in metrics between commits.
-    modify       Modify display properties of data-series plots (has no effect on image-type plots).
+    show        Generate plots from target files or plots definitions from `dvc.yaml` file.
+    diff        Show multiple versions of plot data by plotting them in a single image.
+    modify      Modify display properties of data-series stage plots (has no effect on image-type plots).
```
Can we wrap these texts please? It's a good practice to always sanity check changes [on the review app](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/command-reference/plots):

![image](https://user-images.githubusercontent.com/1477535/178393917-ad6ced17-8ca7-4538-8169-4bda0e8e2168.png)

jorgeorpinel replied:
OK how about this
```suggestion
    show      Generate plots from target files or from `plots`
                definitions in `dvc.yaml`.
    diff      Show multiple versions of a plot by overlaying them
                in a single image.
    modify    Modify display properties of data-series plots
                defined in stages (has no effect on image plots).
```

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -1,6 +1,6 @@
 # plots
 
-A set of commands to visualize and compare _plot metrics_:
+A set of commands to visualize and compare _plot data_:
```
We're not using term "plot data" anywhere else. Should we be specific instead?

```suggestion
A set of commands to visualize and compare data series or images from ML projects:
```
> Needs wrapping

jorgeorpinel commented on content/docs/command-reference/plots/modify.md:
```diff
@@ -1,10 +1,10 @@
 # plots modify
 
-Modify display properties of [plot metrics](/doc/command-reference/plots) files.
+Modify display properties of
+[stage plots](/doc/command-reference/plots#stage-plots) files.
```
```suggestion
Modify display properties of data-series
[plots](/doc/command-reference/plots) defined in <abbr>stages</abbr>.
```

jorgeorpinel commented on content/docs/command-reference/plots/modify.md:
```diff
@@ -16,7 +16,7 @@ usage: dvc plots modify [-h] [-q | -v] [-t <name_or_path>] [-x <field>]
 
 positional arguments:
-  target                Metrics file to set properties to
+  target                Stage plot to set properties to.
```
```suggestion
  target                Plot file to set properties for
```

jorgeorpinel replied:
☝🏼  The description may need a similar update 👇🏼

pared replied:
Done

jorgeorpinel commented on content/docs/command-reference/plots/show.md:
```diff
@@ -143,15 +143,11 @@ file:///Users/usr/src/dvc_plots/index.html
-5,0.9932333,0.02069501801203781,0.9771,0.08009233058886166
-6,0.9945,0.017702101902437668,0.9803,0.07830339228538505
-7,0.9954,0.01396906608727198,0.9802,0.07247738889862157
+epoch,loss,accuracy
```
Seems like the definition, synopsis, description in this file need an update too?

jorgeorpinel commented on content/docs/user-guide/project-structure/pipelines-files.md:
```diff
@@ -459,6 +459,35 @@ validation and auto-completion.
 ⚠️ Note that using the `checkpoint` field in `dvc.yaml` is not compatible with
 `dvc repro`.
 
+## Standalone plots configurations
+
+The list of plots contains one or more user-defined
+[plots](/doc/command-reference/plots#standalone-plots). Here's an example that
+tells DVC that `auc.json` is viable for visualization:
```
```suggestion
## Refusable plots

While [plots](/doc/command-reference/plots#standalone-plots) are ultimately
associated to specific [stages](#stages), you can define one or more of them under a top level
`plots` field in `dvc.yaml` in order to customize them once and reuse across multiple
stages.

The following example tells DVC that output `auc.json` is viable for visualization:
```

dberenbaum replied:
`Refusable` -> `Reusable`

@jorgeorpinel I'm fine with the descriptions, but I'm struggling to understand how the plots are more "reusable" across multiple stages in the top-level field. They enable covering multiple data sources in one plot or using the same data source in multiple plots (maybe that last part is what you mean?).

jorgeorpinel replied:
My understanding here is still not great (didn't have time to read all the motivation in core tickets yet). I thought reusability was the main aspect of this but let's find some other word or go with **Top-level plot definitions**. Or just **Plots** if we're planning to (soft) deprecate stage-level `plots`.

dberenbaum replied:
"Top-level plot definitions" works for now. Let's not let the naming hold up getting this merged because there's a lot of useful content in it.

pared replied:
So I should update the PR and replace `standalone` plots with `top-level definitions`?

dberenbaum replied:
Yup, that sounds good to me.

pared replied:
Done

jorgeorpinel replied:
> They enable covering multiple data sources in one plot or using the same data source in multiple plots

> The main point is that they can use data sourced from different files, maybe we should include that?

Yes, this motivation should be very clear in the docs, probably presented as benefits or features of plots definitions (more than as advantages over stage-level plots, since those may become deprecated soonish -- we shouldn't assume users will know about those as a base case anyway).

I see some motivation is presented in **[Top-level plot definitions](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/user-guide/project-structure/pipelines-files#top-level-plot-definitions)** (currently in the cmd ref). I'll comment there more specifically... ⏳

jorgeorpinel commented on content/docs/user-guide/project-structure/pipelines-files.md:
````diff
@@ -459,6 +459,35 @@ validation and auto-completion.
+    y: tpr
+```
+
+Note that in this example we don't have to specify `auc.json` as a plot output.
+Standalone plots accept any file from repository. Refer to the plots  
+[documentation](/doc/command-reference/plots#syntax) for more information on
+plots syntax.
````
```suggestion
Note that we didn't have to specify `auc.json` as a plot in the stage. In fact,
Top-level plot definitions can use any file in the <abbr>workspace</abbr> or
<abbr>cache</abbr>.

[plots]: /doc/command-reference/plots#standalone-plots
```

pared replied:
I am not sure we should mention cache - "Any file from repository" covers all files - tracked by git,dvc and even those that were just copied into the repo folder and not tracked by anything.

jorgeorpinel replied:
Does this work with non-Git projects?

pared replied:
Yes

jorgeorpinel replied:
Updated suggestion ☝🏼

pared replied:
Updated

jorgeorpinel reviewed (commented):
~~Here's that suggestion I offered for the intro of the schema page. Please review accuracy (I'm still not 100% sure I get the exact motivation for top-level plots).~~

jorgeorpinel commented:
Does the [**Metrics and Plots outputs**](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/user-guide/project-structure/pipelines-files#metrics-and-plots-outputs) section need an update? Maybe split into one for metrics, one for plots? And/or linking to/from some of the new content?

> Sorry, I realize this is a very complex PR. Feel free to split it up it it helps e.g. one for refs, one for the dvc.yaml spec. Up to you

dberenbaum commented:
We can soft-deprecate `plots` inside stages (see https://github.com/iterative/example-repos-dev/pull/117#issuecomment-1175197428). Addressing this throughout the docs should be a separate PR, but if it simplifies this PR to drop references to plots inside stages in the cmd-ref and dvc.yaml description, we can do it here.

dberenbaum commented:
> We can soft-deprecate `plots` inside stages (see [iterative/example-repos-dev#117 (comment)](https://github.com/iterative/example-repos-dev/pull/117#issuecomment-1175197428)). Addressing this throughout the docs should be a separate PR, but if it simplifies this PR to drop references to plots inside stages in the cmd-ref and dvc.yaml description, we can do it here.

Actually, we need to update `dvc plots modify` (see https://github.com/iterative/dvc/issues/7753) to work with top-level plots before we can deprecate stage-level plots, so let's leave it out of this PR please.

jorgeorpinel commented:
@pared there are conflicts here due to a recent PR I made. Unfortunately I can't resolve them on GH from your fork, PTAL when you can. Lmk if you need help solving them and I'll clone your fork to do it locally. Thanks

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -107,7 +98,144 @@ names in the `train` array below:
+<abbr>outputs</abbr> as plot or define plot configuration inside `dvc.yaml`
+under `plots` key.
+
+### Top-level plot definitions
+
+Plots can be defined in `dvc.yaml` under the `plots` key. Unlike
+[stage plots](#stage-plots), these are especially useful when users want to
+compare data from different data sources residing on the same version of the
+project. For example, comparing training versus test results on current branch.
```
> They enable covering **multiple data sources in one plot** or **using the same data source in multiple plots**...
> The main point is that they can **use data sourced from different files**, maybe we should include that?

Also found more specific context in https://github.com/iterative/dvc/issues/7086#issuecomment-985640176:
> 1. iterative/dvc#6316 (single-source segmentation)
> 2. Multiple series per axis (single-source comparison)
> 3. Multiple files within one plot (multi-source composite)

... Any more?

We should expand a bit on the benefits/features of top-level definitions (which may soonish become the only documented place to define plots per https://github.com/iterative/dvc.org/issues/3763). "to compare data from different data sources residing on the same version of the project" doesn't seem to cover all of it.

pared replied:
I think that this one sentence pretty much covers all of it, but I agree that we might not be doing the best job explaining its value here. Let me think how to modify that.

dberenbaum replied:
@jorgeorpinel As you mention above, we shouldn't be comparing to "stage plots" if the "top-level" ones are the default and soon the only supported method. Do we need to explain their unique benefits in that case? Maybe we instead need to explain "stage plots" as a subset that can be defined inside a pipeline stage but are limited to configuring a single output as a plot?

More importantly, should we continue to discuss in this PR or is the rest of it in good enough shape that we could merge and iterate on top?

jorgeorpinel replied:
> > compare data from different data sources
>
> this one sentence pretty much covers all of it

@pared what about the same data source file for 2+ plots?

> Do we need to explain their unique benefits?

@dberenbaum right, I didn't mean the benefits over stage-level plots but rather just the benefits/features of plot definitions (which are top-level).

> explain "stage plots" as a subset

Best not to worry about updating those existing explanations until #3767 is addressed. This PR is already complex.

> should we continue to discuss in this PR or is the rest of it in good enough shape

This question shouldn't a blocker but we should probably make a summary of pending questions and try to get this approved... ⌛

pared replied:
> @pared what about the same data source file for 2+ plots?

You got me here, yes same data source file can also be used

pared commented:
@jorgeorpinel 
rebased and squashed

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -295,9 +432,171 @@ file:///Users/usr/src/dvc_plots/index.html
+
+![](/img/plots_show_confusion_normalized.svg)
+
+### Top-level plots
+
+#### Simple plot definition
+
+Let's get back to the `logs.csv` data:
```
I'm not sure we want many examples if at all in the `plots` ref index page actually. Shouldn't they all be in the sub-refs? Unless a relevant example combines more than one sub-command (in this case I only see `plots show`).

Or if it's about different ways to define plots, that could be summarized into the `dvc.yaml` schema guide.

pared replied:
I guess we should move the examples to `show` `diff`. The syntax is already explained in `Syntax`

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -13,31 +13,22 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify,templates} ...
+projects. Usual plots include AUC curves, loss functions, or confusion matrices,
+for example. Plots are a great alternative to `dvc metrics` when working with
+multi-dimensional performance data. They also help you present and compare
+[[experiments]](https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1024053543) effectively.
```
Oops
```suggestion
[experiments] effectively.
```

jorgeorpinel commented:
Sorry, I think my commits broke the formatting but I can't easily fix it on a fork. Please `git pull`, `yarn && yarn format`, and `git commit -a` @pared . Thanks

jorgeorpinel reviewed (commented):
My only larger concern here is that there's too much info. in the command reference. We should move a good part (probably most) of the new content to the `dvc.yaml` schema "guide". For more specific ideas see the following comments:

- https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1022181003 (use a table for top-level `plots:` syntax)
- https://github.com/iterative/dvc.org/pull/3691#discussion_r918511603 (summarize top-level plots explanation in ref. and move all syntax to dvc.yaml guide)
- https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1045833093 (move `dvc plots` examples to sub-command refs?)

Other less important stuff:

- [x] Provide more motivation for having plot definitions at the top level. See https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1026703393 and https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1041178810
- [ ] Think through `plots` ref examples - https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1045830331
- ~~Update **Metrics and plots** section (dvc.yaml guide)? https://github.com/iterative/dvc.org/pull/3691#issuecomment-1181275871~~
- [x] Does the `plots show` ref need an update too? https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1035149248
- [x] Consistent terminology - https://github.com/iterative/dvc.org/pull/3691#discussion_r918496081
- [x] Mention top-level plots in stage `outs` field desc. https://github.com/iterative/dvc.org/pull/3691#discussion_r922885875
- [x] Remove **Types of metrics** from `metrics` ref. https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1024053543

---
- [ ] See that core texts match usage blocks. See https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1051588011 below.

pared commented on content/docs/command-reference/plots/index.md:
```diff
@@ -202,20 +339,20 @@ this feature to render DVC plots without an Internet connection, below.
+
+#### Tabular data
+
+We'll use tabular data file `logs.csv` for this example:
```
Moved the examples to respective subcommands.

jorgeorpinel replied:
Thanks @pared

pared commented:
> Sorry, I think my commits broke the formatting but I can't easily fix it on a fork. Please git pull, yarn && yarn format, and git commit -a @pared . Thanks

I am sorry for creating this branch on my fork. If we are going to iterate on this I can reopen from main. If we are nearly done we can finish here.

pared commented:
>Provide more motivation for having plot definitions at the top level.

Added some context.

> Think through plots ref examples

Moved to respective subcommands

> Update Metrics and plots section (dvc.yaml guide)?

I don't think we need to. From outputs point of view nothing changes.

dberenbaum commented:
@jorgeorpinel Could you take another look and either approve and put concerns in a new issue, or request specific changes for only those things that you think should block the PR from being merged? It looks like a lot of progress has been made, and I hope we can get this merged soon 🙏 .

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -126,7 +261,7 @@ DVC has the following built-in plot templates:
 - `linear` - basic linear plot including cursor interactivity (default)
 - `simple` - simplest linear template (not interactive); Good base to create
-  [custom templates].
+  [custom template].
 - `scatter` - scatter plot
 - `smooth` - linear plot with LOESS smoothing, see
   [example](/doc/command-reference/plots#example-smooth-plot)
```
Should be plural (1/2)
```suggestion
  [custom templates].
- `scatter` - scatter plot
- `smooth` - linear plot with LOESS smoothing, see
  [example](/doc/command-reference/plots#example-smooth-plot)
```

jorgeorpinel replied:
Shoot I can't easily fix the actual link in line 271... (should be `[custom templates]` there)... ⏳

jorgeorpinel replied:
OK it's time for me to clone this fork 😬 ...

jorgeorpinel replied:
Fixed.

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -13,31 +13,23 @@ usage: dvc plots [-h] [-q | -v] {show,diff,modify,templates} ...
-    diff         Plot differences in metrics between commits.
-    modify       Modify display properties of data-series plots (has no effect on image-type plots).
-    templates    Write built-in plots templates to a directory (.dvc/plots by default).
+    show        Generate plots from target files or from `plots`
+                definitions in `dvc.yaml`.
+    diff        Show multiple versions of a plot by overlaying them
+                in a single image.
+    modify      Modify display properties of data-series plots
+                defined in stages (has no effect on image plots).
+    templates   Write built-in plots templates to a directory
+                (.dvc/plots by default).
```
`plots` sub-command descriptions

pared replied:
This and following suggestions has been used to update https://github.com/iterative/dvc/pull/7972

jorgeorpinel commented on content/docs/command-reference/plots/show.md:
```diff
@@ -12,27 +15,29 @@ usage: dvc plots show [-h] [-q | -v] [-t <name_or_path>] [-x <field>]
 positional arguments:
-  targets               Metrics files to visualize.
-                        Shows all plots by default.
+  targets               Plots files or plot id's from `dvc.yaml` to
+                        visualize. Shows all plots by default.
```
`plots show targets`

jorgeorpinel commented on content/docs/command-reference/plots/modify.md:
```diff
@@ -16,17 +19,18 @@ usage: dvc plots modify [-h] [-q | -v] [-t <name_or_path>] [-x <field>]
 
 positional arguments:
-  target                Metrics file to set properties to
+  target                Plots file to set properties for
+                        (defined at the stage level)
```
`plots modify target`

jorgeorpinel reviewed (commented):
- [ ] (Later) let's check that the core codebase matches these texts.

jorgeorpinel commented:
> take another look and either approve and put concerns in a new issue, or request specific changes 

Well, my main concern is still there. But this PR is approved and I'm not blocking its merge (in fact I [un-blocked](https://github.com/iterative/dvc.org/pull/3691#event-7068015206) it).

_I updated https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1045840235 to reflect the current status -- it could be copied to a follow-up issue if there's no time to address here._

I did partially took this PR over and did lots of small improvements on top of @pared's work. I also did a visual check [of the review app](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/command-reference/plots) and it looks 🆗.

dberenbaum commented:
> @dberenbaum Well, my main concern is still there. But this PR is approved and I'm not blocking it (in fact I [un-blocked](https://github.com/iterative/dvc.org/pull/3691#event-7068015206) it).

Thanks for taking another look! I'm trying to be more explicit about which comments are blocking, so you writing "I'm not blocking it" is helpful.

> _I updated [#3691 (review)](https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1045840235) to reflect the current status -- it could be copied to a follow-up issue if there's no time to address here._

Created https://github.com/iterative/dvc.org/issues/3818.

> I did partially took this PR over and did several improvements on top of @pared's work. I also did a visual check [of the review app](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/command-reference/plots) and it looks 🆗.

Thanks again! I would prefer to merge it if it looks okay and follow up on moving info to the guide. That way, people have somewhere to start referring to use flexible plots.

dberenbaum commented on content/docs/user-guide/project-structure/dvcyaml-files.md:
```diff
@@ -464,6 +476,37 @@ validation and auto-completion.
+The list of `plots` contains one or more user-defined `dvc plots` (paths
+relative to the location of `dvc.yaml`).
+
+This example that makes output `auc.json` viable for visualization, configuring
```
```suggestion
This example makes output `auc.json` viable for visualization, configuring
```

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -105,7 +98,144 @@ names in the `train` array below:
+also used in some top-level definitions, DVC will create separate rendering for
+the stage plots and all definitions using them.
+
+This special type of outputs might come in hand if users want to visually
```
```suggestion
This special type of outputs might come in handy if users want to visually
```

dberenbaum commented on content/docs/command-reference/plots/index.md:
```diff
@@ -105,7 +98,144 @@ names in the `train` array below:
+[stage plots](#stage-plots), these definitions let you overlay plots from
+different data sources, for example training vs. test results (on the current
+project version). Conversely, you can create multiple plots from a single source
+file. You can also any plot file in the project, regardless of whether it's a
+stage outputs. This creates a separation between visualization and outputs.
```
```suggestion
file. You can also use any plot file in the project, regardless of whether it's a
stage outputs. This creates a separation between visualization and outputs.
```

dberenbaum commented on content/docs/command-reference/plots/index.md:
````diff
@@ -105,7 +98,144 @@ names in the `train` array below:
+plots: ...
+```
+
+Every plots has to have its own ID. Configuration, if provided, should be a
+dictionary.
+
+In simplest use case, user can provide file path as the plot ID and don't
+provide configuration at all:
````
```suggestion
Every plot has to have its own ID. Configuration, if provided, should be a
dictionary.

In the simplest use case, a user can provide the file path as the plot ID and not
provide configuration at all:
```

dberenbaum commented:
@pared Can you incorporate the restyled changes? Are you able to merge?

jorgeorpinel commented:
If you (anyone) approves https://github.com/iterative/dvc.org/pull/3819, it should get merged.

dberenbaum commented:
Closing as merged in #3819
